File: Hard/354. Russian Doll Envelopes.py
class Solution:
    def maxEnvelopes(self, envelopes: list) -> int:
        if len(envelopes) < 2:
            return len(envelopes)
        
        envelopes.sort()
        ans = 1
        dp = [1 for _ in range(len(envelopes))]
        
        for i in range(1, len(envelopes)):
            for j in range(i):
                if envelopes[i][0] > envelopes[j][0] and \
                envelopes[i][1] > envelopes[j][1]:
                    dp[i] = max(dp[i], dp[j]+1)
            
            ans = max(ans, dp[i])
        return ans

File: Hard/354. Russian Doll Envelopes.cpp
class Solution {
public:
    int maxEnvelopes(vector<vector<int>>& envelopes) {
        if(envelopes.size() < 2) return envelopes.size();

		sort(envelopes.begin(), envelopes.end());
		int ans = 1;
		int dp[envelopes.size()];
		for(int i = 0; i < envelopes.size(); i++) dp[i] = 1;

		for(int i = 0; i < envelopes.size(); i++){
			for(int j = 0; j < i; j++){
				if(envelopes[i][0] > envelopes[j][0] && envelopes[i][1] > envelopes[j][1]){
					dp[i] = max(dp[i], dp[j]+1);
				}

			}
			ans = max(ans, dp[i]);
		}

		return ans;
    }
};